dask/dataframe/io/orc/core.py

```diff
@@ -0,0 +1,248 @@
+    return new_dd_object(graph, output_name, meta, [None] * (len(parts) + 1))
+
+
+def to_orc(
+    df,
+    path,
+    engine="pyarrow",
+    write_index=True,
+    storage_options=None,
+    compute=True,
+    compute_kwargs=None,
+):
+    """Store Dask.dataframe to ORC files
+
+    Notes
+    -----
+    Each partition will be written to a separate file.
+
+    Parameters
+    ----------
+    df : dask.dataframe.DataFrame
+    path : string or pathlib.Path
+        Destination directory for data.  Prepend with protocol like ``s3://``
+        or ``hdfs://`` for remote data.
+    engine : 'pyarrow' or ORCEngine
+        Parquet library to use. If only one library is installed, it will use
+        that one; if both, it will use 'fastparquet'.
+    write_index : boolean, default True
+        Whether or not to write the index. Defaults to True.
+    storage_options : dict, default None
+        Key/value pairs to be passed on to the file-system backend, if any.
+    compute : bool, default True
+        If True (default) then the result is computed immediately. If False
+        then a ``dask.delayed`` object is returned for future computation.
+    compute_kwargs : dict, default True
+        Options to be passed in to the compute method
+
+    Examples
+    --------
+    >>> df = dd.read_csv(...)  # doctest: +SKIP
+    >>> df.to_orc('/path/to/output/', ...)  # doctest: +SKIP
+
+    See Also
+    --------
+    read_orc: Read ORC data to dask.dataframe
+    """
+
+    # Get engine
+    engine = _get_engine(engine)
+
+    if hasattr(path, "name"):
+        path = stringify_path(path)
+    fs, _, _ = get_fs_token_paths(path, mode="wb", storage_options=storage_options)
+    # Trim any protocol information from the path before forwarding
+    path = fs._strip_protocol(path)
+
+    if not write_index:
+        # Not writing index - might as well drop it
+        df = df.reset_index(drop=True)
+
+    # Use df.npartitions to define file-name list
+    fs.mkdirs(path, exist_ok=True)
+    filenames = [f"part.{i}.orc" for i in range(df.npartitions)]
+
+    # Construct IO graph
+    dsk = {}
+    name = "to-parquet-" + tokenize(
+        df,
+        fs,
+        path,
+        engine,
+        write_index,
+        storage_options,
+    )
+    part_tasks = []
+    for d, filename in enumerate(filenames):
+        dsk[(name, d)] = (
+            apply,
+            engine.write_partition,
+            [
+                (df._name, d),
+                path,
+                fs,
+                filename,
+            ],
+        )
+        part_tasks.append((name, d))
+    dsk[name] = (lambda x: None, part_tasks)
+    graph = HighLevelGraph.from_collections(name, dsk, dependencies=[df])
+    out = Delayed(name, graph)
+
+    # Compute or return future
+    if compute:
+        if compute_kwargs is None:
+            compute_kwargs = dict()
+        out = out.compute(**compute_kwargs)
+    return out
```

Comment:
Similar to https://github.com/dask/dask/pull/7968, it might be good to:

- When `compute=True` use `compute_as_if_collection(DataFrame, ...)`
- When `compute=False` return a `Scalar` instead of a `Delayed`

so that in both cases, users can take advantage of DataFrame optimizations